lls/cache.c

```diff
@@ -0,0 +1,468 @@
+		char ip_buf[cache->key_str_len];
+		char *ip_str;
+		ip_str = cache->ip_str(cache, mod->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, could not get %s map; mod failed\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	RTE_ASSERT(ret >= 0);
+	record = &mod->cache->records[ret];
+
+	if (!is_same_ether_addr(&mod->ha, &record->map.ha)) {
+		ether_addr_copy(&mod->ha, &record->map.ha);
+		changed_ha = true;
+	}
+	if (record->map.port_id != mod->port_id) {
+		record->map.port_id = mod->port_id;
+		changed_port = true;
+	}
+	if (record->map.stale) {
+		record->map.stale = false;
+		changed_stale = true;
+	}
+	record->ts = mod->ts;
+
+	if (changed_ha || changed_port || changed_stale) {
+		lls_update_subscribers(record);
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+	}
+}
+
+unsigned int
+lls_process_reqs(struct lls_config *lls_conf)
+{
+	struct lls_request *reqs[LLS_CACHE_BURST_SIZE];
+	unsigned int count = mb_dequeue_burst(&lls_conf->requests,
+		(void **)reqs, LLS_CACHE_BURST_SIZE);
+	unsigned int i;
+
+	for (i = 0; i < count; i++) {
+		switch (reqs[i]->ty) {
+		case LLS_REQ_HOLD:
+			lls_process_hold(lls_conf, &reqs[i]->u.hold);
+			break;
+		case LLS_REQ_PUT:
+			lls_process_put(lls_conf, &reqs[i]->u.put);
+			break;
+		default:
+			RTE_LOG(ERR, GATEKEEPER,
+				"lls: unrecognized request type (%d)\n",
+				reqs[i]->ty);
+			break;
+		}
+		mb_free_entry(&lls_conf->requests, reqs[i]);
+	}
+
+	return count;
+}
+
+int
+lls_req(enum lls_req_ty ty, void *req_arg)
+{
+	struct lls_config *lls_conf = get_lls_conf();
+	struct lls_request *req = mb_alloc_entry(&lls_conf->requests);
+	int ret;
+
+	if (req == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: allocation for request of type %d failed", ty);
+		return -1;
+	}
+
+	req->ty = ty;
+
+	switch (ty) {
+	case LLS_REQ_HOLD:
+		req->u.hold = *(struct lls_hold *)req_arg;
+		break;
+	case LLS_REQ_PUT:
+		req->u.put = *(struct lls_put *)req_arg;
+		break;
+	default:
+		mb_free_entry(&lls_conf->requests, req);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: unknown request type %d failed", ty);
+		return -1;
+	}
+
+	ret = mb_send_entry(&lls_conf->requests, req);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+void
+lls_cache_scan(struct lls_config *lls_conf, struct lls_cache *cache)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	const uint8_t *ip_be;
+	void *data;
+	struct gatekeeper_if *front = &lls_conf->net->front;
+	struct gatekeeper_if *back = &lls_conf->net->back;
+	time_t now = time(NULL);
+
+	RTE_ASSERT(now >= 0);
+	index = rte_hash_iterate(cache->hash, (void *)&ip_be, &data, &iter);
+	while (index >= 0) {
+		struct lls_record *record = &cache->records[index];
+		uint32_t timeout;
+
+		/*
+		 * If a map is already stale, continue to
+		 * try to resolve it while there's interest.
+		 */
+		if (record->map.stale) {
+			if (record->num_holds > 0)
+				lls_send_request(lls_conf, cache, ip_be, NULL);
+			else
+				lls_del_record(cache, ip_be);
```

Comment:
Does `rte_hash_iterate()` support deletions while scanning? Not all iterators do so.